Use recommended postfix notation for generics

cmeeren opened the pull request:
Fixes #14465

Uses postfix notation for types listed in the [F# code formatting guidelines](https://learn.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#for-types-prefer-prefix-syntax-for-generics-foot-with-some-specific-exceptions):

> 1. For F# Lists, use the postfix form: `int list` rather than `list<int>`.
> 2. For F# Options, use the postfix form: `int option` rather than `option<int>`.
> 3. For F# Value Options, use the postfix form: `int voption` rather than `voption<int>`.
> 4. For F# arrays, use the syntactic name `int[]` rather than `int array` or `array<int>`.
> 5. For Reference Cells, use int ref rather than `ref<int>` or `Ref<int>`.

edgarfgp commented:
> Fixes #14465
> 
> Uses postfix notation for types listed in the [F# code formatting guidelines](https://learn.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#for-types-prefer-prefix-syntax-for-generics-foot-with-some-specific-exceptions):
> 
> > 1. For F# Lists, use the postfix form: `int list` rather than `list<int>`.
> > 2. For F# Options, use the postfix form: `int option` rather than `option<int>`.
> > 3. For F# Value Options, use the postfix form: `int voption` rather than `voption<int>`.
> > 4. For F# arrays, use the syntactic name `int[]` rather than `int array` or `array<int>`.
> > 5. For Reference Cells, use int ref rather than `ref<int>` or `Ref<int>`.

I think number 4. should be  For F# arrays, use the syntactic name `int array` rather than `int[]` or `array<int>`. as https://github.com/dotnet/fsharp/pull/13700 is already in main. cc @vzarytovskii

cmeeren commented:
Note that I just quoted the whole list for completeness. This PR does not tough any array notation, since that is tracked in https://github.com/dotnet/fsharp/issues/14465#issuecomment-1366865317.

edgarfgp commented:
> Note that I simply quoted the whole list for completeness. This PR does not touch any array notation, as per my understanding of [#14465 (comment)](https://github.com/dotnet/fsharp/issues/14465#issuecomment-1366865317).

Thanks for clarifying :)

psfinaki reviewed (changes requested):
Thanks for this, @cmeeren, it's important to be consistent at least with our own guidelines.
I'm afraid we need to update the correspondent `.fs` files, that's why we have all those CI errors.

I am kind of on the fence if this should be an error or not actually. In theory, this is compilable so the error is a bit harsh, but practically I can't see any reason or benefit for such an inconsistency between .fs and .fsi files and it's easy to address both.

Probably should be a warning though.

cmeeren commented:
Ok, do you want me to perform the same replacement in the `.fs` files corresponding to the `.fsi` files I changed?

cmeeren commented:
And importantly, do you want me to update only the required places (corresponding to the `.fsi` updates), or all occurrences in the relevant files?

psfinaki commented:
@cmeeren well yes it would be awesome if you fix the required places, then the CI will pass and we could already merge the PR. But if you have time, feel free to fix some other occurrences as well - you will definitely get extra F# karma points for that :)

cmeeren commented:
I'll do that!

For the record, in case the effect of the signatures aren't clear: The primary reason I am making this change isn't because I am heavily invested in minute stylistic details in FSharp.Core, but because the non-idiomatic notations propagate (via type inference and Intellisense/LineLens etc.) into user code bases.

psfinaki commented:
Thanks a lot! Well, that's anyway beneficial for the F# repo itself as well :)

cmeeren commented:
Hopefully works now. Cleaned up the commits. I have removed all edits from test files, since I am unsure what is safe to change here (some of the tests may be testing the different notations for all I know).

cmeeren commented:
Note that for a while I was thoroughly confused about very many usages of `List<_>` actually being the `System.Collections.Generic` one. You could consider changing that to `ResizeArray` everywhere for clarity, but I can't prioritize making a PR for that at the moment.

psfinaki commented:
@cmeeren thanks for the update, CI is green, @dsyme please reapprove. 

> Note that for a while I was thoroughly confused due to very many usages of List<_> actually being the System.Collections.Generic one. You could consider changing that to the ResizeArray alias everywhere for clarity.

I am on the fence with this actually, but feel free to create a ticket and elaborate a bit, someone can pick it up then.

Note I created a ticket for an analyzer to force the conventions, I think that would be useful and expected by C# users.

dsyme commented:
> You could consider changing that to the ResizeArray alias everywhere for clarity. 

I agree we should do this in our own code

dsyme commented on src/Compiler/Driver/CompilerConfig.fs:
```diff
@@ -300,7 +300,7 @@ type ImportedAssembly =
         mutable TypeProviders: Tainted<ITypeProvider> list
 #endif
-        FSharpOptimizationData: Microsoft.FSharp.Control.Lazy<Option<Optimizer.LazyModuleInfo>>
+        FSharpOptimizationData: Microsoft.FSharp.Control.Lazy<Optimizer.LazyModuleInfo option>
```
FWIW this should just be `Lazy` - not sure why the full namespace is quoted - can clean up separately
